kube-runtime/src/finalizer.rs

```diff
@@ -0,0 +1,146 @@
+    Api,
+};
+use serde::{de::DeserializeOwned, Serialize};
+use snafu::{ResultExt, Snafu};
+use std::{error::Error as StdError, fmt::Debug};
+
+#[derive(Debug, Snafu)]
+pub enum Error<ApplyErr, CleanupErr>
+where
+    ApplyErr: StdError + 'static,
+    CleanupErr: StdError + 'static,
+{
+    #[snafu(display("failed to apply object: {}", source))]
+    ApplyFailed { source: ApplyErr },
+    #[snafu(display("failed to clean up object: {}", source))]
+    CleanupFailed { source: CleanupErr },
+    #[snafu(display("failed to add finalizer: {}", source))]
+    AddFinalizer { source: kube::Error },
+    #[snafu(display("failed to remove finalizer: {}", source))]
+    RemoveFinalizer { source: kube::Error },
+}
+
+/// Reconcile an object in a way that requires cleanup before an object can be deleted. It does this by
+/// managing a [`ObjectMeta::finalizers`] entry, which prevents the object from being deleted before the
+/// cleanup is done.
+///
+/// In typical usage, if you use `finalizer` then it should be the only top-level "action"
+/// in your [`applier`]/[`Controller`]'s `reconcile` function.
+///
+/// # Expected Flow
+///
+/// 1. User creates object
+/// 2. Reconciler sees object
+/// 3. `finalizer` adds `finalizer_name` to [`ObjectMeta::finalizers`]
+/// 4. Reconciler sees updated object
+/// 5. `finalizer` runs `apply`
+/// 6. User updates object
+/// 7. Reconciler sees updated object
+/// 8. `finalizer` runs `apply`
+/// 9. User deletes object
+/// 10. Reconciler sees deleting object
+/// 11. `finalizer` runs `cleanup`
+/// 12. `finalizer` removes `finalizer_name` from [`ObjectMeta::finalizers`]
+/// 13. Kubernetes sees that all [`ObjectMeta::finalizers`] are gone and finally deletes the object
+///
+/// # Guarantees
+///
+/// If `apply` is ever started then `cleanup` must succeed before the Kubernetes object deletion completes.
+///
+/// # Assumptions
+///
+/// `finalizer_name` must be unique among the controllers interacting with the object
+///
+/// `apply` and `cleanup` must both be idempotent, and tolerate being executed several times (even if previously cancelled).
+///
+/// `cleanup` must tolerate `apply` never having ran at all, or never having succeeded. Keep in mind that
+/// even infallible `.await`s are cancellation points.
+///
+/// # Caveats
+///
+/// Object deletes will get stuck while the controller is not running, or if `cleanup` fails for some reason.
+///
+/// # Errors
+///
+/// `apply` and `cleanup` are both fallible, their errors are passed through as [`Error::ApplyFailed`]
+/// and [`Error::CleanupFailed`], respectively.
+///
+/// In addition, adding and removing the finalizer itself may fail. In particular, this may be because of
+/// network errors, or because another `finalizer` was updated in the meantime on the same object.
+pub async fn finalizer<K, ApplyFut, CleanupFut>(
+    api: &Api<K>,
+    finalizer_name: &str,
+    obj: K,
+    apply: impl FnOnce(K) -> ApplyFut,
+    cleanup: impl FnOnce(K) -> CleanupFut,
```

Comment:
Taking two closures like this means that Rust doesn't understand that it "should conceptually be" safe to move the same value into both closures, because only one will be executed (although it isn't actually, since that would still lead to double `drop`s).

A few alternate approaches:

1. A single callback, that takes the "mode" (apply vs cleanup) as an `enum`
   - \+ Allows safely sharing moved values (avoiding needing new `clone`s and maybe even `Arc`s)
   - \- Slightly awkward, need to name the mode just to pattern-match on it
   - \- Might give the impression that it's a good idea to share logic between the modes
   - \- Prevents different `Error` types for `apply` and `cleanup`
2. Return the "mode" instead of running any callbacks at all
   - \- User needs to manually remove the finalizer after cleanup

Reply:
I have a slight feeling that the single callback might make more sense. if we are requiring that both cleanup/apply are idempotent anyway, perhaps encouraging re-use isn't so bad? i am sure you could still do metrics/tracing easier if it was a single fn.

Reply:
If we had a single fn, we could almost wrap root reconcile without too many surprising changes.If so, I feel like an I.e. i feel like an ideal entrypoint would look like `Controller::run`:

```rust
Controller::new(cmgs, ...)
     .owns(cms, ...)
     .run_with_finalizer(reconcile, error_policy, context)
```

and here `reconcile` in this case would maybe have a different parameters (because it's not a standard reconciler), and maybe it also takes the suggested `enum FinalizeMode`.

Though I don't have a complete mental picture here. Not sure if that makes sense.

Reply:
(although i think my main problem with two fns is that they are two indistinguishable comma separated closures passed to a fn call. If we had to builder ctors for them, we could infer their intent without consulting kube-runtime's source, and it would probably be better.)

Reply:
> although i think my main problem with two fns is that they are two indistinguishable comma separated closures passed to a fn call.

Yeah that makes sense. Guess I was a bit blinded by Rust-Analyzer's parameter name inlays.

Reply:
Okay, tried merging them into an `Event` enum now